achievements.lua

```diff
@@ -77,37 +77,129 @@
+
+	local function split_all_after_first(str, char)
+		local char_pos = str:find(char, 1, true)
+		local res = nil
+		if char_pos ~= nil then
+			status, res = pcall(function() return { str:sub(0, char_pos - 1), str:sub(char_pos + 1, -1) } end)
+			if status then
+				return res[1], res[2]
+			end
+		end
+		return str, ""
+	end
+
+	-- drop everything after '+' (as is normal in semver)
+	ver, _ = split_all_after_first(ver, '+')
+	-- also drop everything after '-' (as parsing that is massive overkill for the intended use)
+	ver, _ = split_all_after_first(ver, '-')
+
+	-- what is left should be a normal version number
+	local result = {}
+	while ver ~= nil and ver ~= "" do
+		local num_str, next = split_all_after_first(ver, '.')
+		if num_str ~= nil and num_str ~= "" then
+			local num = nil
+			status, num = pcall(function() return tonumber(num_str) end)
+			if not status then
+				num = -1
+			end
+			table.insert(result, num)
+		end
+		ver = next
+	end
+
+	return result
+end
+
+local function compare_version_strings(version_a, version_b)
+	local ver_a = parse_version_string(version_a)
+	local ver_b = parse_version_string(version_b)
+	local longest = math.max(#ver_a, #ver_b)
+	for i_pos = 1, longest, 1 do
+		ver_a[i_pos] = ver_a[i_pos] or 0
+		ver_b[i_pos] = ver_b[i_pos] or 0
+		if ver_a[i_pos] < ver_b[i_pos] then
+			return -1
+		end
+		if ver_a[i_pos] > ver_b[i_pos] then
+			return 1
+		end
+	end
+	return 0
+end
+
 local function get_achievement_folder_root_path(gameID)
 	if type(gameID) ~= "string" then
 		error("bad argument #1: expected string, got " .. type(gameID), 2)
 	end
-	local root = string.format("/Shared/%s/", gameID)
+	local root = string.format(default_shared_achievement_folder .. "%s/", gameID)
 	return root
 end
 local function get_achievement_data_file_path(gameID)
 	if type(gameID) ~= "string" then
 		error("bad argument #1: expected string, got " .. type(gameID), 2)
 	end
 	local root = get_achievement_folder_root_path(gameID)
-	return root .. "Achievements.json"
+	return root .. default_achievement_file_name
+end
+local function get_shared_images_path(gameID)
+	if type(gameID) ~= "string" then
+		error("bad argument #1: expected string, got " .. type(gameID), 2)
+	end
+	local root = get_achievement_folder_root_path(gameID)
+	return root .. default_shared_images_subfolder
+end
+local function get_shared_images_updated_file_path(gameID)
+	if type(gameID) ~= "string" then
+		error("bad argument #1: expected string, got " .. type(gameID), 2)
+	end
+	local folder = get_shared_images_path(gameID)
+	return folder .. default_shared_images_updated_file
 end
-
--- local achievement_folder = root_folder .. playdate.metadata.bundleID .. "/"
 
 local metadata <const> = playdate.metadata
+local gfx <const> = playdate.graphics
 
 ---@diagnostic disable-next-line: lowercase-global
 achievements = {
-	version = "prototype 0.1"
+	specversion = "0.1+prototype",
+	libversion = "0.2-alpha+prototype",
+
+	onUnconfigured = error,
```

Comment:
I'll change this, since I thought about it a bit more and I think you're right, but in case you wanted to know my original reasoning;

I was thinking along the lines of what'll probably get tested by inexperienced developers. The achievement-names are all strings after all.

On the one hand we have 'fail fast', which is a good guiding principle. On the other, crashing a game because a very uncommon and untested achievement got misspelled by an un-experienced dev might be a problem as well.

Reply:
Done!